Rename #expires_now and add #http_cache_never

nateberkopec opened the pull request:
See https://github.com/rails/rails/pull/19556. cc @sgrif who was assigned to that PR.

nateberkopec committed:
expires_now -> always_revalidate || http_cache_never

The original documentation on expires_now was slightly misleading, as in most
contexts "cache" is the same thing as "store". In HTTP caching, this is not
the case, as something which is "not cached" may still be stored and revalidated
on each request. Some users may have though the expires_now method prevented
browsers and intermediate caches from storing responses, which it would not.

In addition, expires_now was a bit of a confusing name for adding a cache-control
directive of "no-cache". By renaming "expires_now" to "always_revalidate", we are
revealing the intention behind the method.

Also, provided is an "http_cache_never" method intended to be the anti-counterpart
to "http_cache_forever". It adds a "no-store" directive, which will prevent the
response from being stored *or* cached by any caches.

nateberkopec commented:
This method is still very poorly named, but I'm not pr0 enough to know how Rails might handle deprecating/changing the name of such an old method.

pixeltrix commented:
@nateberkopec deprecation of an existing method is straightforward - just wrap a call to the new method from the old method with a warning, e.g:

``` ruby
def old_method(*args)
  ActiveSupport::Deprecation.warn("old_method is deprecated and will be removed in Rails 5.1. Please use new method instead")
  new_method(*args)
end
```

Then in Rails 5.1 we just delete the old method. :smile:

nateberkopec commented:
@pixeltrix Cool, updated with what I think we should actually do. Commit message:

>The original documentation on expires_now was slightly misleading, as in most
contexts "cache" is the same thing as "store". In HTTP caching, this is not
the case, as something which is "not cached" may still be stored and revalidated
on each request. Some users may have though the expires_now method prevented
browsers and intermediate caches from storing responses, which it would not.
>
> In addition, expires_now was a bit of a confusing name for adding a cache-control
directive of "no-cache". By renaming "expires_now" to "always_revalidate", we are
revealing the intention behind the method.
>
> Also, provided is an "http_cache_never" method intended to be the anti-counterpart
to "http_cache_forever". It adds a "no-store" directive, which will prevent the
response from being stored *or* cached by any caches.

maclover7 commented:
Can you add some `assert_deprecated` tests?

rafaelfranca commented on actionpack/lib/action_controller/metal/conditional_get.rb:
```diff
@@ -208,12 +208,31 @@ def expires_in(seconds, options = {})
+    # resources that need to be revalidated on each request, but may sometimes
+    # be unchanged.
+    def always_revalidate
       response.cache_control.replace(:no_cache => true)
```
new hash syntax

rafaelfranca commented on actionpack/lib/action_controller/metal/conditional_get.rb:
```diff
@@ -208,12 +208,31 @@ def expires_in(seconds, options = {})
+    # Use for sensitive or private information, but be aware that intermediates
+    # or eavesdroppers may ignore this directive and privacy cannot be guaranteed.
+    def http_cache_never
+      response.cache_control.replace(:no_store => true)
```
New hash syntax

rafaelfranca commented on actionpack/test/controller/render_test.rb:
```diff
@@ -113,14 +113,19 @@ def conditional_hello_with_expires_in_with_public_with_more_keys_old_syntax
     response.headers['Cache-Control'] = 'no-transform'
-    expires_now
+    always_revalidate
+    render :action => 'hello_world'
```
new hash syntax

nateberkopec commented:
Fixed hash syntax and added deprecation test.

I'm confused on why my tests are failing though - something is replace the `no-store` directive with `private`, which doesn't even make sense, as they are not the same thing. Can someone give me a lead as to what's doing this?

Also changed the behavior of `always_revalidate` to add a directive rather than replace all of them. This is because it may be useful to specify a header of `private no-cache` - only allow browsers to store this response, and they must always revalidate it.

Pinging @matthewd who seems to ~~also be a sadist~~ enjoy reading W3C specs as much as I do, let me know what you think about this change.

nateberkopec commented:
Also, I'm wondering if anyone thinks that `always_revalidate` is sort of a silly method now - you can just pass `no_cache` as an option to `expires_in`.
